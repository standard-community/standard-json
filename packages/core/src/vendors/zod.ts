import type { JSONSchema7 } from "json-schema";
import { zodToJsonSchema } from "zod-to-json-schema";
import type * as z from "zod";

export const toJsonSchema = (
  schema: z.ZodType,
  options?: Record<string, unknown>,
) => zodToJsonSchema(schema, options) as JSONSchema7;
