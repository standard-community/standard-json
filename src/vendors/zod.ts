import type { ZodTypeAny } from "zod/v3";
import type { $ZodType } from "zod/v4/core";
import { errorMessageWrapper, type ToJsonSchemaFn } from "./utils.js";

export default async function getToJsonSchemaFn(): Promise<ToJsonSchemaFn> {
  return async (schema, options) => {
    let handler: ToJsonSchemaFn;

    // https://zod.dev/library-authors?id=how-to-support-zod-and-zod-mini-simultaneously#how-to-support-zod-3-and-zod-4-simultaneously
    if ("_zod" in (schema as $ZodType | ZodTypeAny)) {
      try {
        handler = await import("zod/v4/core").then(
          (mod) => mod.toJSONSchema as ToJsonSchemaFn,
        );
      } catch {
        throw new Error(errorMessageWrapper('Missing dependencies "zod v4".'));
      }
    } else {
      try {
        handler = await import("zod-to-json-schema").then(
          (mod) => mod.zodToJsonSchema as ToJsonSchemaFn,
        );
      } catch {
        throw new Error(
          errorMessageWrapper('Missing dependencies "zod-to-json-schema".'),
        );
      }
    }

    return handler(schema, options);
  };
}
