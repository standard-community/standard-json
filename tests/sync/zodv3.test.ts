import { describe, expect, it } from "vitest";
import { type ZodTypeAny, z } from "zod";
import { zodToJsonSchema } from "zod-to-json-schema";
import { loadVendor, toJsonSchema } from "~/index.js";
import type { ToJsonSchemaFn } from "~/vendors/utils.js";

describe("sync check", () => {
  it("zodv3", () => {
    // Loading the vendor
    loadVendor("zod", ((schema, options) =>
      zodToJsonSchema(schema as ZodTypeAny, options)) as ToJsonSchemaFn);

    const zod3Schema = z
      .object({
        myString: z.string(),
        myUnion: z.union([z.number(), z.boolean()]),
      })
      .describe("My neat object schema");

    const zod3JsonSchema = toJsonSchema.sync(zod3Schema);
    expect(zod3JsonSchema).toMatchSnapshot();
  });
});
