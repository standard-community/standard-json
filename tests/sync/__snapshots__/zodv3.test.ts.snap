// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sync check > zodv3 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "description": "My neat object schema",
  "properties": {
    "myString": {
      "type": "string",
    },
    "myUnion": {
      "type": [
        "number",
        "boolean",
      ],
    },
  },
  "required": [
    "myString",
    "myUnion",
  ],
  "type": "object",
}
`;
